include/boost/windows_sspi/stream.hpp

```diff
@@ -33,14 +33,67 @@ namespace boost {
 namespace windows_sspi {
 
 // TODO: Move away from this file
-template <typename NextLayer, typename ConstBufferSequence> struct async_write_impl : boost::asio::coroutine {
+template <typename NextLayer>
+struct async_handshake_impl : boost::asio::coroutine {
+  async_handshake_impl(NextLayer& next_layer, detail::sspi_impl& sspi_impl)
+    : m_next_layer(next_layer)
+    , m_sspi_impl(sspi_impl) {
+  }
+
+  template <typename Self>
+  void operator()(Self& self, boost::system::error_code ec = {}, std::size_t length = 0) {
+    BOOST_ASIO_CORO_REENTER(*this) {
```

Comment:
You can save yourself some error checking code repetition by putting this above the re-enter:
```
if (ec) 
  return self.complete(ec);
```

Reply:
That's a very good point. Thanks.